geps/gep-3155/index.md

```diff
@@ -0,0 +1,205 @@
+
+### SANs on BackendTLSPolicy
+
+This change enables the certificate to have a different identity than the SNI
+(both are currently tied to the hostname field). This is particularly useful
+when using SPIFFE, which relies on URI Subject Names which are not valid SNIs 
+as per https://www.rfc-editor.org/rfc/rfc6066.html#section-3.
+
+In such case either connection properties or an arbitrary SNI, like cluster-local 
+service name could be used for certificate selection, while the identity validation
+will be done based on SubjectAltNames field.
+
+**1. Add a new `SubjectAltNames` field to `BackendTLSPolicyValidation`**
+
+```go
+type BackendTLSPolicyValidation struct {
+  // SubjectAltNames contains one or more Subject Alternative Names.
+  // When specified, at least one of certificate's Subject Alternate Names MUST 
+  // match at least one of the specified SubjectAltNames.
+  // +kubebuilder:validation:MaxItems=5
+  SubjectAltNames []SubjectAltName `json:"subjectAltNames,omitempty"`
+}
+
+type SubjectAltName struct {
+  // Hostname contains Subject Alternative Name specified in DNS name format. It is
+  // mutually exclusive with Hostname. At least one of the fields must be specified.
+  Hostname v1.PreciseHostname `json:"hostname"`
+
+  // URI contains Subject Alternative Name specified in URI format. It is mutually
+  // exclusive with Hostname. At least one of the fields must be specified.
+  URI string `json:"uri,omitempty"`
+}
+```
+
+**2. Modify Spec for `BackendTLSPolicyValidation` `Hostname`**
+
+Before:
+```go
+	// 2. Hostname MUST be used for authentication and MUST match the certificate
+	//    served by the matching backend.
+```
+
+After:
+```go
+	// 2. Only when SubjectAltNames is not specified, Hostname MUST be used for 
+    //    authentication and MUST match the certificate served by the matching
+    //    backend.
+    // 3. If HostName is unspecified, SNI header will be omitted.
+```
+
+### Allow per-service mTLS settings BackendTLSPolicy
+
+Gateway level TLS configuration already includes an `options` field. This has
+been helpful for implementation-specific TLS configurations, or simply features
+that have not made it to the core API yet. It would be similarly useful to have
+an identical field on BackendTLSPolicy.
+
+```go
+type BackendTLSPolicySpec struct {
+	// Options are a list of key/value pairs to enable extended TLS
+	// configuration for each implementation. For example, configuring the
+	// minimum TLS version or supported cipher suites.
+	//
+	// A set of common keys MAY be defined by the API in the future. To avoid
+	// any ambiguity, implementation-specific definitions MUST use
+	// domain-prefixed names, such as `example.com/my-custom-option`.
+	// Un-prefixed names are reserved for key names defined by Gateway API.
+	//
+	// Support: Implementation-specific
+	//
+	// +optional
+	// +kubebuilder:validation:MaxProperties=16
+	Options map[AnnotationKey]AnnotationValue `json:"options,omitempty"`
+}
+```
```

Comment:
I understand that these escape hatches are handy, but as a maintainer I'm concerned that they have two effects:

* They promote the creation of non-portable configuration, which is against the purpose of this API
* They encourage users to use them and not go through the (yes, sometimes arduous) process of giving feedback and building features into upstream.

The fact that you're saying this is useful and I have no idea how this field is actually being used in the field illustrates exactly what I mean. If you can get everything you need out of the spec by using a free-text escape hatch like this, why would you ever do the work to come and ask for proper typed, structured fields?

To put this another way: everytime I see a `map[string]string`, I think of Ingress annotations and the terrible situation that has ended up in. So I, personally, need to see a pretty big benefit to justify these risks. I'd like to see some concrete use cases here as well that illustrate the need for this.

Reply:
I believe this will be primarily used for experimental features, before they are actually merged. First two I expect are:
- constraints like min-tls-version
- configuration for proprietary / experimental managed TLS modes